src/rust/engine/src/externs/dep_inference.rs

```diff
@@ -59,12 +59,20 @@ impl PyInferenceMetadata {
         )))
     }
 
-    fn __richcmp__(&self, other: &Self, op: CompareOp, py: Python) -> PyObject {
-        match op {
-            CompareOp::Eq => (self == other).into_py(py),
-            CompareOp::Ne => (self != other).into_py(py),
+    fn __richcmp__(&self, other: &Self, op: CompareOp, py: Python) -> PyResult<PyObject> {
+        Ok(match op {
+            CompareOp::Eq => (self == other)
+                .into_pyobject(py)?
+                .to_owned()
+                .into_any()
+                .unbind(),
```

Comment:
This seems like a lot of hoop-jumping to "convert" a Rust bool into a Python bool.  Shouldn't this function (even before this change) return `PyResult<bool>` and then the formal conversion is a lot simpler?

Reply:
The default branch of the match for `py.NotImplemented` is what is causing the rest of mess since all three branches have to return the same type. If the function returns `PyResult<bool>` then it cannot return `NotImplemented`.

Reply:
I have a solution in mind for a custom enum (with `True`, `False`, and `NotImplemented` variants) which would be convertible from `Option<bool>` and convert to the correct Python-side types. But that would be a follow-on PR.

Reply:
Ugh, I would have expected we could raise `NotImplementedError` via the PyErr in the PyResult, but I guess technically you are supposed to return the `NotImplemented` singleton in comparison operators, and not raise. So this is correct.